proposed/2023/NuGetAudit-without-nuget.org-as-packageSource.md

```diff
@@ -0,0 +1,217 @@
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+#### auditSources in nuget.config
+
+*NuGet.Config* files have a new section `<auditSources>` where the URL(s) of any audit sources can be specified, which will be used only to download a VDB, and will not be used for downloading packages.
+Like `<packageSources>`, `<auditSources>` can contain zero, one, or more sources, and supports `<clear />`.
+When there is at least one source in `auditSources`, NuGet should no longer use `packageSources` to get a VDB.
+
+For example, consider a developer at Contoso using a company-internal source as a single package source, and using nuget.org as an audit source:
+
+```xml
+<configuration>
+  <packageSources>
+    <clear />
+    <add key="contoso" value="https://internal.corp.contoso.com/packages/nuget/index.json" >
+  </packageSources>
+  <auditSources>
+    <clear />
+    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" >
+  </auditSources>
+</configuration>
+```
+
+Any URL specified in the `auditSources` section must implement the [NuGet Server API V3 protocol](https://learn.microsoft.com/nuget/api/overview).
+However, the server will only need to implement the [`VulnerabilityInfo` resource](https://learn.microsoft.com/en-us/nuget/api/vulnerability-info) (and service index), and can avoid implementing all the other resources that are required to be used as a package source.
+
+NuGet will use same semantics for `auditSources` as it already has with `packageSources`.
+This means that `auditSources` are accumulated from other nuget.config files, and `<clear />` can be used to remove them.
+`auditSources` will also follow `packageSources` with regards to insecure http warnings, errors, and `allowInsecureConnections` configurations.
+Sources that need credentials will use `packageSourceCredentials`, using the `key` as the lookup, just as `packageSources` does.
+This means that credentials can be re-used by using the same key for a source that is defined in both `packageSources` and `auditSources`.
+
+#### Restore summary
+
+When restore is run at normal or detailed verbosity, NuGet outputs detected nuget.config files and package sources (feeds) used.
+Audit sources should be added.
+
+```diff
+ NuGet Config files used:
+     C:\Users\zivkan\AppData\Roaming\NuGet\NuGet.Config
+
+ Feeds used:
+     https://api.nuget.org/v3/index.json
++
++Audit sources used:
++    https://api.nuget.org/v3/index.json
+```
+
+When `auditSources` do not have any configured sources, replay all of the package sources again under the audit sources used section, to make it explicit to customers that these sources are being used.
+
+If `NuGetAudit` is disabled for all projects in the solution, the audit sources used section of the summary should be omitted.
```

Comment:
This is something I only came up with as I was updating the spec. We didn't discuss this in the meeting. What do others think?

I'm not actually sure about the technical feasibility of checking all projects to see if NuGetAudit is enabled/disabled. But if a customer has disabled Audit for all their projects, it might look a bit weird if the restore summary still claims an audit source was used.

Should we have a message "NuGet Audit is disabled for all projects in the solution" instead?

Note that there's another related feature request:

* https://github.com/NuGet/Client.Engineering/issues/2340

I do not want this "audit sources used" mini-feature to become scope creep into implementing the entirety of all NuGetAudit related restore summary. But, this might be a quick win that can be changed when the "full" audit restore summary feature is designed & implemented.

Reply:
Agreed. We should not show this experience unless users opt into the functionality. This higher level switch should influence the behavior down the line for other experiences.

We do not need a message here if the functionality is disabled. That's the user's choice and flexibility they desire (to not see those messages)